sources/web/datalab/backupUtility.ts

```diff
@@ -0,0 +1,100 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="../../../externs/ts/node/node.d.ts" />
+/// <reference path="common.d.ts" />
+
+import childProcess = require('child_process');
+import logging = require('./logging');
+import path = require('path');
+import fs = require('fs');
+
+var appSettings: common.Settings;
+var backupIntervalVar: number = null;
+var backup_log_path: string = '';
+
+// check for backups every 10 minutes so we can retry failed backups
+const BACKUP_INTERVAL: number = 10*60*60;
+
+/**
+ * Last successful backup timestamps are stored in a file to avoid unnecessary backups,
+ * and also to compare current time against last time a successful backup ran
+ */
+function readBackupLog() {
+  try {
+    return JSON.parse(fs.readFileSync(backup_log_path, 'utf8'));
+  } catch (err) {
+    return {};
+  }
+}
+function writeBackupLog(log_history: Object) {
+  return fs.writeFileSync(backup_log_path, JSON.stringify(log_history), 'utf8');
+}
+
+/**
+ * Run the GCSbackup shell script and call the callback function with its exit code
+ */
+function runBackup(tag: string, callback: Function) {
+  var cmd = childProcess.spawn('/bin/bash', ['/datalab/GCSbackup.sh',
+                               '-t', tag,
+                               '-p', '/content/datalab',
+                               '-l', '/datalab/.backup_log.txt']);
+  cmd.on('close', (code: number) => {
+    if (code > 0) {
+      logging.getLogger().error('WARNING: Backup script with tag ' + tag + ' failed with code: ' + code);
+    } else {
+      callback();
+    }
+  });
+}
+
+/**
+ * Start a periodic backup process by calling the GCSbackup shell script
+ * every hour, day, and week. Compare against the last time each backup
+ * ran to catch up if we missed any backups (e.g. because Datalab wasn't running)
+ */
+export function startBackup(settings: common.Settings) {
+  appSettings = settings;
+  backup_log_path = path.join(appSettings.datalabRoot, '/datalab/.backup_history');
+
+  if (!backupIntervalVar && settings.enableAutoGCSBackups) {
+    var num_hourly = settings.numHourlyBackups;
+    var num_daily = settings.numDailyBackups;
+    var num_weekly = settings.numWeeklyBackups;
+
+    backupIntervalVar = global.setInterval(() => {
+      var log_history = readBackupLog();
+      // test hourly backup
+      if (!log_history.lastHourlyRun || (Date.now()-log_history.lastHourlyRun)/1000 > 60*60) {
+        runBackup('hourly', () => {
+          writeBackupLog(log_history);
+          log_history.lastHourlyRun = Date.now();
+        });
+      }
+      if (!log_history.lastDailyRun || (Date.now()-log_history.lastDailyRun)/1000 > 60*60*24) {
+        runBackup('daily', () => {
```

Comment:
Here's my reasoning why I think all three should be separate:
- Simplifies the implementation since each run only has to look for previous runs with the same tag.
- If one backup fails it doesn't affect the others.
- If backups are not complete because VM was stopped, these backups can go out of sync, and that should be fine.